Bottom: 8e761860f23e84f22834d20d35d3daa930782c33
Top:    280c60267ad5a68f4cc06fd1990d852594424baa
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2012-12-22 16:17:49 +0000

find.c: Avoid adding any rat-lines at all when AndRats is false.

Previously we were adding some rat-lines when searching for physical
connections. We should stop at the physical->rat boundary, not the
opposite end of the ratline.


---

diff --git a/src/find.c b/src/find.c
index dfa77db..bea37c2 100644
--- a/src/find.c
+++ b/src/find.c
@@ -285,10 +285,10 @@ static bool LookupLOConnectionsToPVList (bool);
 static bool LookupLOConnectionsToLOList (bool);
 static bool LookupPVConnectionsToLOList (bool);
 static bool LookupPVConnectionsToPVList (void);
-static bool LookupLOConnectionsToLine (LineType *, Cardinal, bool);
-static bool LookupLOConnectionsToPad (PadType *, Cardinal);
-static bool LookupLOConnectionsToPolygon (PolygonType *, Cardinal);
-static bool LookupLOConnectionsToArc (ArcType *, Cardinal);
+static bool LookupLOConnectionsToLine (LineType *, Cardinal, bool, bool);
+static bool LookupLOConnectionsToPad (PadType *, Cardinal, bool);
+static bool LookupLOConnectionsToPolygon (PolygonType *, Cardinal, bool);
+static bool LookupLOConnectionsToArc (ArcType *, Cardinal, bool);
 static bool LookupLOConnectionsToRatEnd (PointType *, Cardinal);
 static bool IsRatPointOnLineEnd (PointType *, LineType *);
 static bool ArcArcIntersect (ArcType *, ArcType *);
@@ -859,21 +859,21 @@ LookupLOConnectionsToLOList (bool AndRats)
                   position = &lineposition[layer];
                   for (; *position < LineList[layer].Number; (*position)++)
                     if (LookupLOConnectionsToLine
-                        (LINELIST_ENTRY (layer, *position), group, true))
+                        (LINELIST_ENTRY (layer, *position), group, true, AndRats))
                       return (true);
 
                   /* try all new arcs */
                   position = &arcposition[layer];
                   for (; *position < ArcList[layer].Number; (*position)++)
                     if (LookupLOConnectionsToArc
-                        (ARCLIST_ENTRY (layer, *position), group))
+                        (ARCLIST_ENTRY (layer, *position), group, AndRats))
                       return (true);
 
                   /* try all new polygons */
                   position = &polyposition[layer];
                   for (; *position < PolygonList[layer].Number; (*position)++)
                     if (LookupLOConnectionsToPolygon
-                        (POLYGONLIST_ENTRY (layer, *position), group))
+                        (POLYGONLIST_ENTRY (layer, *position), group, AndRats))
                       return (true);
                 }
               else
@@ -889,7 +889,7 @@ LookupLOConnectionsToLOList (bool AndRats)
                   position = &padposition[layer];
                   for (; *position < PadList[layer].Number; (*position)++)
                     if (LookupLOConnectionsToPad
-                        (PADLIST_ENTRY (layer, *position), group))
+                        (PADLIST_ENTRY (layer, *position), group, AndRats))
                       return (true);
                 }
             }
@@ -1727,7 +1727,7 @@ LOCtoArcPad_callback (const BoxType * b, void *cl)
  * Xij means Xj at arc i
  */
 static bool
-LookupLOConnectionsToArc (ArcType *Arc, Cardinal LayerGroup)
+LookupLOConnectionsToArc (ArcType *Arc, Cardinal LayerGroup, bool AndRats)
 {
   Cardinal entry;
   struct lo_info info;
@@ -1861,7 +1861,7 @@ LOCtoLinePad_callback (const BoxType * b, void *cl)
  */
 static bool
 LookupLOConnectionsToLine (LineType *Line, Cardinal LayerGroup,
-                           bool PolysTo)
+                           bool PolysTo, bool AndRats)
 {
   Cardinal entry;
   struct lo_info info;
@@ -1871,12 +1871,15 @@ LookupLOConnectionsToLine (LineType *Line, Cardinal LayerGroup,
   info.line = Line;
   search_box = expand_bounds ((BoxType *)info.line);
 
-  /* add the new rat lines */
-  if (setjmp (info.env) == 0)
-    r_search (PCB->Data->rat_tree, &search_box, NULL,
-              LOCtoLineRat_callback, &info);
-  else
-    return true;
+  if (AndRats)
+    {
+      /* add the new rat lines */
+      if (setjmp (info.env) == 0)
+        r_search (PCB->Data->rat_tree, &search_box, NULL,
+                  LOCtoLineRat_callback, &info);
+      else
+        return true;
+    }
 
   /* loop over all layers of the group */
   for (entry = 0; entry < PCB->LayerGroups.Number[LayerGroup]; entry++)
@@ -2135,14 +2138,14 @@ LOCtoPadPad_callback (const BoxType * b, void *cl)
  * layergroup. All found connections are added to the list
  */
 static bool
-LookupLOConnectionsToPad (PadType *Pad, Cardinal LayerGroup)
+LookupLOConnectionsToPad (PadType *Pad, Cardinal LayerGroup, bool AndRats)
 {
   Cardinal entry;
   struct lo_info info;
   BoxType search_box;
 
   if (!TEST_FLAG (SQUAREFLAG, Pad))
-    return (LookupLOConnectionsToLine ((LineType *) Pad, LayerGroup, false));
+    return (LookupLOConnectionsToLine ((LineType *) Pad, LayerGroup, false, AndRats));
 
   info.pad = Pad;
   search_box = expand_bounds ((BoxType *)info.pad);
@@ -2150,11 +2153,14 @@ LookupLOConnectionsToPad (PadType *Pad, Cardinal LayerGroup)
   /* add the new rat lines */
   info.layer = LayerGroup;
 
-  if (setjmp (info.env) == 0)
-    r_search (PCB->Data->rat_tree, &search_box, NULL,
-              LOCtoPadRat_callback, &info);
-  else
-    return true;
+  if (AndRats)
+    {
+      if (setjmp (info.env) == 0)
+        r_search (PCB->Data->rat_tree, &search_box, NULL,
+                  LOCtoPadRat_callback, &info);
+      else
+        return true;
+    }
 
   /* loop over all layers of the group */
   for (entry = 0; entry < PCB->LayerGroups.Number[LayerGroup]; entry++)
@@ -2274,7 +2280,7 @@ LOCtoPolyRat_callback (const BoxType * b, void *cl)
  * on the given layergroup. All found connections are added to the list
  */
 static bool
-LookupLOConnectionsToPolygon (PolygonType *Polygon, Cardinal LayerGroup)
+LookupLOConnectionsToPolygon (PolygonType *Polygon, Cardinal LayerGroup, bool AndRats)
 {
   Cardinal entry;
   struct lo_info info;
@@ -2286,12 +2292,16 @@ LookupLOConnectionsToPolygon (PolygonType *Polygon, Cardinal LayerGroup)
   search_box = expand_bounds ((BoxType *)info.polygon);
 
   info.layer = LayerGroup;
+
   /* check rats */
-  if (setjmp (info.env) == 0)
-    r_search (PCB->Data->rat_tree, &search_box, NULL,
-              LOCtoPolyRat_callback, &info);
-  else
-    return true;
+  if (AndRats)
+    {
+      if (setjmp (info.env) == 0)
+        r_search (PCB->Data->rat_tree, &search_box, NULL,
+                  LOCtoPolyRat_callback, &info);
+      else
+        return true;
+    }
 
 /* loop over all layers of the group */
   for (entry = 0; entry < PCB->LayerGroups.Number[LayerGroup]; entry++)
